File: anagram.ts
// Algorithm pattern: FREQUENZY LOOP
// Time complexity: O(N)
// Space complexity: O(N)

const isAnagram = (str1: string, str2: string): boolean => {
  // Check if two strings have same amount of same characters: "iceman" = "cinema"
  // create arrays for both strings
  const str1Arr = str1.split("");
  const str2Arr = str2.split("");
  // Create character object from first array and store its characters as keys with initial value of 1
  const str1Obj = {};
  // in case character object contains character already, increment its value by 1
  for (let ch1 in str1Arr) {
    str1Arr[ch1] in str1Obj
      ? (str1Obj[str1Arr[ch1]] += 1)
      : (str1Obj[str1Arr[ch1]] = 1);
  }
  // Loop thorough second string array characters to see if they exist in character object
  // if exists, decrement it from the character object
  for (let ch2 in str2Arr) {
    str2Arr[ch2] in str1Obj ? (str1Obj[str1Arr[ch2]] -= 1) : null;
  }
  // Create a new Set with unique values of the character object
  // create an array from the object values
  // if all values in the set are 0, then its size is 1, so return true
  const str1ObjArr = Array.from(Object["values"]<number>(str1Obj));
  const str1Set = new Set<number>(str1ObjArr);
  return str1Set.size === 1;
};

isAnagram("icemxn", "cinema");

File: anagram.py
# Algorithm pattern: FREQUENZY LOOP
# Time complexity: O(N)
# Space complexity: O(N)

def is_anagram(str_1, str_2):
    # See if both the strings contain an equal amount of the same characters, if not, return False
    if len(str_1) != len(str_2):
        return False
    # Create dictionary and save characters there as keys with initial value 1
    str_1_dict = {}
    # save first string's characters to array
    str_1_list = list(str_1)
    # in dictionary increment value by 1 if key already exists
    for ch_1 in str_1_list:
        if ch_1 not in str_1_dict:
            str_1_dict.update({ch_1: 1})
        else:
            str_1_dict[ch_1] += 1
    # Compare second string characters and values to first string and return False if they don't match
    # make array of second string
    str_2_list = list(str_2)
    # if character from second string array is found on dict, decrement it
    for ch_2 in str_2_list:
        if ch_2 in str_1_dict:
            str_1_dict[ch_2] -= 1
    # In dict if key contains value larger than 1, return False
    # make a set of dict values, and if they are all 0, then length of set is 1, else return False
    val_set = set(str_1_dict.values())
    return True if len(val_set) == 1 else False


def main():
    print(is_anagram("iceman", "cinema"))


if __name__ == '__main__':
    main()
